### PR TITLE
Chart section ordering

### PR DESCRIPTION
**Pull request by KhadijaMahanga**
## Description

- [x] Add chart section ordering

Fixes [#169497311](https://www.pivotaltracker.com/story/show/169497311)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Desktop Screenshots

Peek 2019-11-18 18-02

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: wp-content/plugins/hurumap/db.php [hunk 1]
```diff
@@ -100,6 +100,45 @@ function migrate() {
 
     }
 
+    if ($version < 4) {
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/db.php:101]: Version here is the same as as in https://github.com/TakwimuAfrica/Dashboard/pull/21/files#diff-199986d3b01e1279a89a1dd8f387e586R101. If I've run that migration, my understanding `$version` will now be `4` which means this migration **won't** run.

**Review by kilemensi — COMMENTED** [review]
Still can't run this @KhadijaMahanga .. deleted `node_modules`, reinstalled and rebuilt the plugin but no luck. As I've described below, I think one of the issues is the database versioning we are using.

File: wp-content/plugins/hurumap/src/App.js [hunk 2]
```diff
@@ -270,8 +271,9 @@ function App() {
                     <Button
                       className={classes.button}
                       onClick={() =>
-                        arrayHelper.insert(0, {
+                        arrayHelper.push({
                           id: shortid.generate(),
+                          order: 1000,
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/src/App.js]: No need to use 1000. Lets use `arrayHelper.length` i.e. goes to the back of the list.

We can then use `index` as `order`:
1. if ` == 0`, disable up,
2. if ` == arrayHelper.length - 1` disable down,
3. Up/down means swapping elements at respective indexes, and
4. Removing just means removing that element from array.

After any of the above change, we just save the indexes as order.

File: wp-content/plugins/hurumap/src/App.js [hunk 3]
```diff
@@ -289,65 +291,111 @@ function App() {
...
-                                );
-                                if (chart) {
-                                  handleUpdateHurumapChart(form, chart, {
-                                    section: section.id
-                                  });
-                                } else {
-                                  chart = form.values.flourishCharts.find(
+                    <Grid container direction="column" spacing={2}>
+                      {form.values.sections
+                        .sort((a, b) => (a.order > b.order ? 1 : -1))
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/src/App.js]: Just to make sure there is no unnecessary reordering, `a.order >=  b.order` i.e. if they're both 10, `a` will maintain it's position above `b`

File: wp-content/plugins/hurumap/src/App.js [hunk 4]
```diff
@@ -289,65 +291,111 @@ function App() {
...
+                                  } else if (
+                                    position === 'down' &&
+                                    q < form.values.sections.length - 1
+                                  ) {
+                                    swapSection = form.values.sections[q + 1];
+                                    arrayHelper.swap(q, q + 1);
+                                    updateOrCreateChartsSection([
+                                      Object.assign(section, { order: q + 1 }),
+                                      Object.assign(swapSection, { order: q })
+                                    ]);
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/src/App.js]: Lets move this into a function outside `JSX` (same for `onDelete` below)

**Review by kilemensi — COMMENTED** [review]
:+1: @KhadijaMahanga ... Deleting the DB docker instance seemed to do the trick. I can at least test the PR.

As we discussed, the charts also need to be order-able within a section since we'll no longer have the `charts.json` file that we can use.

File: wp-content/plugins/hurumap/db.php (hunk 1 above)
Comment (KhadijaMahanga) [reply wp-content/plugins/hurumap/db.php:101]: @kilemensi , yes version is the same. Maybe it was what was causing the error on local. But I was hoping this two PRs will be pushed at the same time to production. So no need to add another version

Comment (kilemensi) [reply wp-content/plugins/hurumap/db.php:101]: Here be dragons!

File: wp-content/plugins/hurumap/src/App.js (hunk 4 above)
Comment (KhadijaMahanga) [reply wp-content/plugins/hurumap/src/App.js]: :+1:

File: wp-content/plugins/hurumap/src/App.js (hunk 3 above)
Comment (KhadijaMahanga) [reply wp-content/plugins/hurumap/src/App.js]: :+1:

File: wp-content/plugins/hurumap/db.php [hunk 5]
```diff
@@ -100,8 +100,20 @@ function migrate() {
...
 
-        $sql = "CREATE TABLE `{$wpdb->base_prefix}hurumap_charts`(
+        $sql = "CREATE TABLE `{$wpdb->base_prefix}chart_sections`(
+            `id`         varchar(45) NOT NULL,
+            `name`       varchar(255) NOT NULL ,
+            `description` varchar(255) NOT NULL ,
+            `order`         int NOT NULL,
+            `published` tinyint NOT NULL DEFAULT 0,
+            `created_at` datetime NOT NULL DEFAULT NOW() ON UPDATE NOW(),
+            `updated_at` datetime NOT NULL DEFAULT NOW(),
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/db.php]: I think these are reversed, `created_at` should never be updated _after_ the row is created. `updated_at` is the one we want to update on every row update.

File: wp-content/plugins/hurumap/src/App.js [hunk 6]
```diff
@@ -270,84 +298,90 @@ function App() {
...
+                    <Grid container direction="column" spacing={2}>
+                      {form.values.sections
+                        .sort((a, b) => (a.order >= b.order ? 1 : -1))
+                        .map((section, q) => (
+                          <Grid key={section.id} item xs={12} md={6}>
+                            <React.Suspense fallback={<div>Loading...</div>}>
+                              <ChartsSection
+                                section={Object.assign(section, { order: q })}
+                                sectionsCount={form.values.sections.length}
+                                onMove={position => {
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/src/App.js]: ```onMove={movement => {```

File: wp-content/plugins/hurumap/src/App.js [hunk 7]
```diff
@@ -270,84 +298,90 @@ function App() {
...
+                                  if (position === 'up') {
```
Comment (kilemensi) [inline wp-content/plugins/hurumap/src/App.js]: Can't the movement check and array swapping be done in the `handleXXXChartsSection` functions? i.e. include `movement` and `arrayHelper` as parameters to these function.

File: wp-content/plugins/hurumap/db.php (hunk 1 above)
Comment (kilemensi) [reply wp-content/plugins/hurumap/db.php:101]: This still bugs me @KhadijaMahanga ... Lets add a story to track it.

**Review by kilemensi — APPROVED** [review]
:rocket: @KhadijaMahanga 

----------------------------------------

1. I experienced the delay/hanging when creating HURUmap charts... I think we're making background api calls that hang the UI,
2. I guess we're going to sort charts in a different PR, right?

File: wp-content/plugins/hurumap/src/App.js (hunk 7 above)
Comment (KhadijaMahanga) [reply wp-content/plugins/hurumap/src/App.js]: :+1:

File: wp-content/plugins/hurumap/src/App.js (hunk 6 above)
Comment (KhadijaMahanga) [reply wp-content/plugins/hurumap/src/App.js]: :100:

**Comment by KhadijaMahanga** [comment]
@kilemensi added a bug story here https://www.pivotaltracker.com/story/show/169947872

**Pull request merged**